oauth2_http/java/com/google/auth/oauth2/IdTokenCredentials.java

```diff
@@ -0,0 +1,199 @@
+ *     .setIdTokenProvider(saCreds)
+ *     .setTargetAudience(targetAudience).build();
+ * 
+ * // for ComputeEngineCredentials
+ * ComputeEngineCredentials caCreds = ComputeEngineCredentials.create();
+ * IdTokenCredentials tokenCredential = IdTokenCredentials.newBuilder()
+ *     .setIdTokenProvider(caCreds)
+ *     .setTargetAudience(targetAudience)
+ *     .setOptions(Arrays.asList(ComputeEngineCredentials.ID_TOKEN_FORMAT_FULL))
+ *     .build();
+ * 
+ * // for ImpersonatedCredentials
+ * ImpersonatedCredentials imCreds = ImpersonatedCredentials.create(saCreds,
+ *     "impersonated-account@project.iam.gserviceaccount.com", null,
+ *     Arrays.asList("https://www.googleapis.com/auth/cloud-platform"), 300);
+ * IdTokenCredentials tokenCredential = IdTokenCredentials.newBuilder()
+ *     .setIdTokenProvider(imCreds)
+ *     .setTargetAudience(targetAudience)
+ *     .setOptions(Arrays.asList(ImpersonatedCredentials.INCLUDE_EMAIL))
+ *     .build();
+ *
+ * // Use the IdTokenCredential in an authorized transport
+ * GenericUrl genericUrl = new GenericUrl("https://example.com");
+ * HttpCredentialsAdapter adapter = new HttpCredentialsAdapter(tokenCredential);
+ * HttpTransport transport = new NetHttpTransport();
+ * HttpRequest request = transport.createRequestFactory(adapter).buildGetRequest(genericUrl);
+ * HttpResponse response = request.execute();
+ *
+ * // Print the token, expiration and the audience
+ * System.out.println(tokenCredential.getIdToken().getTokenValue());
+ * System.out.println(tokenCredential.getIdToken().getJsonWebSignature().getPayload().getAudienceAsList());
+ * System.out.println(tokenCredential.getIdToken().getJsonWebSignature().getPayload().getExpirationTimeSeconds());
+ * </pre>
+ */
+@Beta
+public class IdTokenCredentials extends OAuth2Credentials {
+
+  private static final long serialVersionUID = -2133257318957588431L;
+
+  private IdTokenProvider idTokenProvider;
+  private String targetAudience;
+  private List<IdTokenProvider.Option> options;
+
+
+  private IdTokenCredentials(Builder builder) {
+    this.idTokenProvider = builder.getIdTokenProvider();
+    this.targetAudience = builder.getTargetAudience();
+    this.options = builder.getOptions();
+  }
+
+  @Override
+  public AccessToken refreshAccessToken() throws IOException {
+    return this.idTokenProvider.idTokenWithAudience(targetAudience, options);
+  }
+
+  public IdToken getIdToken() {
+    return (IdToken) getAccessToken();
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(idTokenProvider);
+  }
+
+  @Override
+  public String toString() {
+    return MoreObjects.toStringHelper(this).toString();
+  }
+
+  @Override
+  public boolean equals(Object obj) {
+    if (!(obj instanceof IdTokenCredentials)) {
+      return false;
+    }
+    IdTokenCredentials other = (IdTokenCredentials) obj;
+    return Objects.equals(this.idTokenProvider, other.idTokenProvider)
+        && Objects.equals(this.targetAudience, other.targetAudience);
```

Comment:
Bug: equals depends on targetAudience but hashCode does not

Reply:
ok, so added in like this

```java
  @Override
  public int hashCode() {
    return Objects.hash(idTokenProvider, targetAudience);
  }
```